core/src/main/java/org/apache/iceberg/util/ExpireSnapshotUtil.java

```diff
@@ -0,0 +1,394 @@
+   * may refer to files which are no longer accessible from any valid snapshot and do not effect
+   * the current table.
+   *
+   * For this we need to look through
+   *   1. Snapshots which have not expired but contain manifests from expired snapshots
+   *   2. Snapshots which have expired and contain manifests referring to now orphaned files
+   *
+   * @param snapshotChanges       Information about which Snapshots are still valid and which have been expired
+   * @param original              The table metadata from before the snapshot expiration
+   * @param io                    FileIO for reading manifest info
+   * @return Wrapper around which manifests contain references to possibly abandoned files
+   */
+  public static ManifestExpirationChanges determineManifestChangesFromSnapshotExpiration(
+      SnapshotExpirationChanges snapshotChanges, TableMetadata original, FileIO io) {
+
+    AncestorInfo ancestorInfo = getAncestorInfo(original);
+
+    Set<Snapshot> currentSnapshots = snapshotChanges.getValidSnapshots();
+    Set<Snapshot> expiredSnapshots = snapshotChanges.getExpiredSnapshots();
+    Set<Long> validIds = snapshotChanges.getValidSnapshotIds();
+    Set<Long> expiredIds = snapshotChanges.getExpiredSnapshotIds();
+    Set<Long> ancestorIds = ancestorInfo.getAncestorIds();
+
+    //Snapshots which are not expired but refer to manifests from expired snapshots
+    Set<ManifestFile> validManifests = getValidManifests(currentSnapshots, io);
+    Set<ManifestFile> manifestsToScan = findValidManifestsInExpiredSnapshots(validManifests, ancestorInfo, validIds);
+
+    //Snapshots which are expired and do not effect the current table
+    List<Snapshot> snapshotsNotInTableState = findSnapshotsNotInTableState(expiredSnapshots, ancestorInfo);
+    ManifestExpirationChanges manifestExpirationChanges =
+        findExpiredManifestsInUnusedSnapshots(snapshotsNotInTableState, validManifests, ancestorIds, expiredIds, io);
+
+    manifestExpirationChanges.manifestsToScan().addAll(manifestsToScan);
+    return manifestExpirationChanges;
+  }
+
+  /**
+   * Compares the Snapshots from the two TableMetadata objects and identifies the snapshots
+   * still in use and those no longer in use
+   * @param current Metadata from a table after an expiration of snapshots
+   * @param original Metadata from the table before expiration of snapshots
+   * @return Information about which Snapshots have Expired and which are Still Valid
+   */
+  public static SnapshotExpirationChanges getExpiredSnapshots(TableMetadata current, TableMetadata original) {
+
+    Set<Snapshot> validSnapshots = Sets.newHashSet();
+    for (Snapshot snapshot : current.snapshots()) {
+      validSnapshots.add(snapshot);
+    }
+
+    Set<Snapshot> expiredSnapshots = Sets.newHashSet();
+    for (Snapshot snapshot : original.snapshots()) {
+      if (!validSnapshots.contains(snapshot)) {
```

Comment:
Seems like we don't implement `hashcode` and `equals` for `BaseSnapshot`.  Won't that be a problem here?

Reply:
We are comparing the same objects in memory here, but we could do a more complicated hash and equals if we want. Originally this was just a match made on the SnapshotId but since everything is dealing with the same references I thought this would be a bit clearer.